File: backend/internal/contrib/contrib.go
package contrib

import (
	"encoding/json"
	"net/http"
	"strconv"
	"sync"
	"time"
)

const apiURL = "https://github-contributions-api.jogruber.de/v4/valtterisa?y=last"

type Day struct {
	Date  string `json:"date"`
	Count int    `json:"count"`
	Level int    `json:"level"`
	Tip   string `json:"-"`
}

type Graph struct {
	Days    []Day
	Caption string
}

var (
	client   = &http.Client{Timeout: 8 * time.Second}
	mu       sync.Mutex
	cached   Graph
	cachedAt time.Time
)

func Get() Graph {
	mu.Lock()
	defer mu.Unlock()
	if cached.Days != nil && time.Since(cachedAt) < 24*time.Hour {
		return cached
	}
	days := fetch()
	if days != nil {
		cached = Graph{Days: days, Caption: caption(days)}
		cachedAt = time.Now()
	}
	return cached
}

func fetch() []Day {
	res, err := client.Get(apiURL)
	if err != nil {
		return nil
	}
	defer res.Body.Close()
	if res.StatusCode != http.StatusOK {
		return nil
	}
	var data struct {
		Contributions []Day `json:"contributions"`
	}
	if json.NewDecoder(res.Body).Decode(&data) != nil {
		return nil
	}
	for i, d := range data.Contributions {
		data.Contributions[i].Tip = tip(d)
	}
	return lastMonths(data.Contributions, 12)
}

func lastMonths(days []Day, months int) []Day {
	cutoff := time.Now().UTC().AddDate(0, -months, 0)
	for cutoff.Weekday() != time.Sunday {
		cutoff = cutoff.AddDate(0, 0, -1)
	}
	cutoffStr := cutoff.Format("2006-01-02")
	for i, d := range days {
		if d.Date >= cutoffStr {
			return days[i:]
		}
	}
	return nil
}

func caption(days []Day) string {
	total := 0
	for _, d := range days {
		total += d.Count
	}
	start, end := days[0].Date[:4], days[len(days)-1].Date[:4]
	year := start
	if end != start {
		year = start + "–" + end + " ytd"
	}
	return strconv.Itoa(total) + " contributions in " + year
}

func tip(d Day) string {
	date := d.Date
	if t, err := time.Parse("2006-01-02", d.Date); err == nil {
		date = t.Format("Jan 2, 2006")
	}
	switch d.Count {
	case 0:
		return "No contributions on " + date
	case 1:
		return "1 contribution on " + date
	default:
		return strconv.Itoa(d.Count) + " contributions on " + date
	}
}

File: src/lib/contrib.ts
import {
  isJsonArray,
  isJsonObject,
  isNumber,
  isString,
  readJsonValue,
  type JsonObject,
  type JsonValue,
} from "./json";

export type ContributionDay = {
  date: string;
  count: number;
  level: number;
  tip: string;
};

export type ContributionGraph = {
  days: ContributionDay[];
  caption: string;
};

type RawContributionDay = {
  date: string;
  count: number;
  level: number;
};

const apiUrl =
  "https://github-contributions-api.jogruber.de/v4/valtterisa?y=last";

const cacheTtlMs = 8 * 60 * 60 * 1000;

let cachedGraph: ContributionGraph | null = null;
let cachedAt = 0;

function formatTip(day: Pick<RawContributionDay, "date" | "count">): string {
  const date = new Date(`${day.date}T00:00:00Z`);
  const formatted = date.toLocaleDateString("en-US", {
    month: "short",
    day: "numeric",
    year: "numeric",
    timeZone: "UTC",
  });

  if (day.count === 0) {
    return `No contributions on ${formatted}`;
  }
  if (day.count === 1) {
    return "1 contribution on " + formatted;
  }
  return `${day.count} contributions on ${formatted}`;
}

function parseContributionDay(value: JsonObject): RawContributionDay | null {
  if (
    !isString(value.date) ||
    !isNumber(value.count) ||
    !isNumber(value.level)
  ) {
    return null;
  }

  return {
    date: value.date,
    count: value.count,
    level: value.level,
  };
}

function parseContributionsResponse(value: JsonValue | null): RawContributionDay[] {
  if (!isJsonObject(value) || !isJsonArray(value.contributions)) {
    return [];
  }

  return value.contributions
    .map((entry) => (isJsonObject(entry) ? parseContributionDay(entry) : null))
    .filter((day): day is RawContributionDay => day !== null);
}

function lastMonths(days: RawContributionDay[], months: number): RawContributionDay[] {
  const cutoff = new Date();
  cutoff.setUTCMonth(cutoff.getUTCMonth() - months);
  while (cutoff.getUTCDay() !== 0) {
    cutoff.setUTCDate(cutoff.getUTCDate() - 1);
  }

  const cutoffStr = cutoff.toISOString().slice(0, 10);
  const index = days.findIndex((day) => day.date >= cutoffStr);
  return index === -1 ? [] : days.slice(index);
}

function buildCaption(days: ContributionDay[]): string {
  const total = days.reduce((sum, day) => sum + day.count, 0);
  const startYear = days[0]?.date.slice(0, 4) ?? "";
  const endYear = days[days.length - 1]?.date.slice(0, 4) ?? "";
  const year =
    endYear !== startYear ? `${startYear}–${endYear} ytd` : startYear;
  return `${total} contributions in ${year}`;
}

async function fetchContributions(): Promise<ContributionDay[] | null> {
  try {
    const response = await fetch(apiUrl, {
      cache: "no-store",
      signal: AbortSignal.timeout(8000),
    });
    if (!response.ok) {
      return null;
    }

    const data = await readJsonValue(response);
    const trimmed = lastMonths(parseContributionsResponse(data), 12);

    return trimmed.map((day) => ({
      ...day,
      tip: formatTip(day),
    }));
  } catch {
    return null;
  }
}

export async function getContributionGraph(): Promise<ContributionGraph> {
  const now = Date.now();
  if (cachedGraph && now - cachedAt < cacheTtlMs) {
    return cachedGraph;
  }

  const days = await fetchContributions();
  if (days && days.length > 0) {
    cachedGraph = {
      days,
      caption: buildCaption(days),
    };
    cachedAt = now;
    return cachedGraph;
  }

  return cachedGraph ?? { days: [], caption: "" };
}

export function parseContributionGraph(
  value: JsonValue | null,
): ContributionGraph | null {
  if (!isJsonObject(value) || !isJsonArray(value.days) || !isString(value.caption)) {
    return null;
  }

  const days: ContributionDay[] = [];
  for (const entry of value.days) {
    if (!isJsonObject(entry)) {
      return null;
    }
    if (
      !isString(entry.date) ||
      !isNumber(entry.count) ||
      !isNumber(entry.level) ||
      !isString(entry.tip)
    ) {
      return null;
    }
    days.push({
      date: entry.date,
      count: entry.count,
      level: entry.level,
      tip: entry.tip,
    });
  }

  return { days, caption: value.caption };
}

export async function fetchContributionGraphClient(): Promise<ContributionGraph> {
  const response = await fetch("/api/contrib");
  if (!response.ok) {
    throw new Error("Failed to load contribution graph");
  }

  const raw = (await response.json()) as JsonValue;
  const parsed = parseContributionGraph(raw);
  if (!parsed || parsed.days.length === 0) {
    throw new Error("Invalid contribution graph");
  }

  return parsed;
}
